Ful 2540: Password Instant Validation & Strength Indicator

bsudy opened the pull request.

Commit:
FUL-2540: Added initial version of password strength indicator logic

Commit:
Merge branch 'master' into FUL-2540

Commit:
FUL-2540: Added Universal Module Definition (https://github.com/umdjs/umd/blob/master/templates/jqueryPlugin.js)

Commit:
FUL-2540: Added grunt build with uglification and minification

Commit:
FUL-2540: Design adjusments

Commit:
FUL-2540: Change strength namingChanged colour scheme

Commit:
FUL-2540: Changed coloursintroduced variables in stylesheet

Commit:
FUL-2540: Code clean.Added better option namingAdded documentation

Commit:
FUL-2540: Amended documentation

Wattos commented on jquery.password_field.js:
```diff
@@ -64,6 +152,115 @@
+
+            $fragment.insertAfter($this);
+
+            $this.on('input', function(event) {
```
careful, this does not work correctly on IE9 which does not fire the input event on DELETE or Backspace. See https://developer.mozilla.org/en-US/docs/Web/Events/input

Maybe lets just use keyup

Wattos commented on jquery.password_field.js:
```diff
@@ -64,6 +152,115 @@
+                var password = $this.val();
+
+                var hasDigit = /\d/.test(password);
+                var hasLower = /[a-z]/.test(password);
```
This will not work for unicode characters, e.g. ü which is a valid lower case character. Maybe we should use something along the lines of:

```
function isLowerCase(ch) {
    return ch == ch.toLowerCase() && ch != ch.toUpperCase();
}

function isUpperCase(ch) {
    return ch == ch.toUpperCase() && ch != ch.toLowerCase();
}

hasUpper = _.any(password, isLowerCase)
hasLower = _.any(password, isUpperCase)

```

Wattos commented on jquery.password_field.js:
```diff
@@ -64,6 +152,115 @@
+                var strength = 'na';
+                if (password) {
+                    switch (result.score) {
+                        case 4:
```
I think having a map is probably cleaner in this case?

```
map = {
   4: 'strong'
   3: 'good',
   2: 'medium',
   1: 'weak'
}

```

Wattos commented on jquery.password_field.js:
```diff
@@ -64,6 +152,115 @@
+            },
+            validity: {
+                header: 'Your password must have',
+                pwd_length: '8 or more characters',
```
let's not use pwd. We should spell it out. We should also use camelCase in javascript instead of snake_case

Wattos commented on jquery.password_field.sass:
```diff
@@ -12,4 +12,90 @@ span.visiblity-toggle-wrapper
+        display: none
+
+    li:before
+        font: normal normal normal 14px/1 FontAwesome
```
what is `14px/1` ?

Wattos commented on jquery.password_field.sass:
```diff
@@ -12,4 +12,90 @@ span.visiblity-toggle-wrapper
+
+$weak-color: #DE404D
+$okay-color: #DE404D
+$medium-color: #FCE147
```
Medium color is not readable

Commit:
FUL-2540: Changed plugin to use invalid/good/strong measures instead of separate validity and strength meter

Commit:
FUL-2540: Code clean and fixes
Fixed lowercase - uppercase detection making it utf compatible
Changed 'input' event to 'keyup' in order to get it working on ie9
Fixed naming issues

Commit:
FUL-2540: Added class names to make login more straightforward

Commit:
FUL-2540: Added correct 'good' (orange) color

Wattos commented:
LGTM. Thanks! :+1:
